### ISSUE TITLE
Performance gain after adding HasValue check to nullable type comparison 

### ISSUE THREAD

**Issue by odk0706** [dotnet/efcore#18782]
For some reason, we get a big performance gain when we add an additional HasValue check to nullable long type comparison. Is this behavior on purpose or can this be labeled as a bug?

Does this maybe have something to do with UseRelationalNulls?

### Steps to reproduce

``` C#
// WITH HASVALUE CHECK

 b.VirtualBookings =
                    await _context.BookingVirtuals.AsNoTracking().Where(e => e.ORIGINALBOOKINGID.HasValue && e.ORIGINALBOOKINGID == b.Booking.BOOKINGID).ToListAsync() ??
                    new List<BookingVirtual>();

// WITHOUT HASVALUE CHECK

 b.VirtualBookings =
                    await _context.BookingVirtuals.AsNoTracking().Where(e => e.ORIGINALBOOKINGID == b.Booking.BOOKINGID).ToListAsync() ??
                    new List<BookingVirtual>();

// ORIGINALBOOKINGID type = long?
// BOOKINGID type = long
```

``` SQL
-- WITH HASVALUE CHECK
exec sp_executesql N'SELECT [a].[BOOKINGID],  /* REMOVED COLUMN NAMES */
FROM [AllBookings] AS [a]
WHERE ([a].[booking_type] = 1) AND ([a].[ORIGINALBOOKINGID] IS NOT NULL AND 
((([a].[ORIGINALBOOKINGID] = @__b_Booking_BOOKINGID_0) AND ([a].[ORIGINALBOOKINGID] IS NOT NULL AND @__b_Booking_BOOKINGID_0 IS NOT NULL)) OR 
([a].[ORIGINALBOOKINGID] IS NULL AND @__b_Booking_BOOKINGID_0 IS NULL)))',N'@__b_Booking_BOOKINGID_0 bigint',@__b_Booking_BOOKINGID_0=6743540

-- WITHOUT HASVALUE
exec sp_executesql N'SELECT [a].[BOOKINGID], /* REMOVED COLUMN NAMES */
FROM [AllBookings] AS [a]
WHERE ([a].[booking_type] = 1) AND ((([a].[ORIGINALBOOKINGID] = @__b_Booking_BOOKINGID_0) AND 
([a].[ORIGINALBOOKINGID] IS NOT NULL AND @__b_Booking_BOOKINGID_0 IS NOT NULL)) OR 
([a].[ORIGINALBOOKINGID] IS NULL AND @__b_Booking_BOOKINGID_0 IS NULL))',N'@__b_Booking_BOOKINGID_0 bigint',@__b_Booking_BOOKINGID_0=6743540
```

### Test results:

The query was executed on a SQL Server 2016 (13.0.5081)  database with around 6.7 million records in the queried table.

#### With HasValue:

(4 rows affected)
Table 'AllBookings'. Scan count 1, logical reads 19, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

(1 row affected)

 SQL Server Execution Times:
   CPU time = 0 ms,  elapsed time = 3 ms.
 SQL Server Execution Times:
   CPU time = 0 ms,  elapsed time = 3 ms.
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.

#### Without HasValue:

(4 rows affected)
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'AllBookings'. Scan count 5, logical reads 19023, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.
Table 'Worktable'. Scan count 0, logical reads 0, physical reads 0, read-ahead reads 0, lob logical reads 0, lob physical reads 0, lob read-ahead reads 0.

(1 row affected)

 SQL Server Execution Times:
   CPU time = 1688 ms,  elapsed time = 476 ms.
 SQL Server Execution Times:
   CPU time = 1688 ms,  elapsed time = 476 ms.
SQL Server parse and compile time: 
   CPU time = 0 ms, elapsed time = 0 ms.
 SQL Server Execution Times:
   CPU time = 0 ms,  elapsed time = 0 ms.

### Further technical details

EF Core version: 3.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.0
Operating system: Windows 10 Pro 1903 x64 
IDE: Visual Studio 2019 16.3.0

**Comment by ErikEJ** [comment]
I think this is fixed in 3.1

**Comment by ajcvickers** [comment]
@maumar Is this already fixed?

**Comment by smitpatel** [comment]
HasValue check will short-circuit early in case of null hence performance gain. I believe in 3.1 we added that logic in our pipeline too that if you write HasValue like that, the column in the right side of && will always be non-null if it reached there.

**Comment by IGx89** [comment]
Just ran into this bug ourselves. Query conditions in EF Core 2.2 were:
```
WHERE [c].[PartyId] = @__partyId_0
```
and in EF Core 3.0 are:
```
WHERE (([c].[PartyId] = @__partyId_0) AND ([c].[PartyId] IS NOT NULL AND @__partyId_0 IS NOT NULL)) OR ([c].[PartyId] IS NULL AND @__partyId_0 IS NULL)
```
Our Azure SQL Server database that was running very comfortable at 20 DTUs in production started maxing out 200 DTUs after our deploy completely due to that query change. This is a hard issue to catch in non-prod environments unless you're able to simulate production-level loads -- due to that I'd personally consider this a critical bug.

**Comment by maumar** [comment]
@odk0706 @IGx89 this issue has been fixed in 3.1, we removed unnecessary null checks so the performance around null comparisons should be comparable to 2.2. Original issue: #17543

**Comment by IGx89** [comment]
Mainly including my experience in case there are plans for a 3.0.1 bugfix release. With 3.1 apparently tracking for December though I'm guessing that's basically going to be your bugfix release.

**Comment by maumar** [comment]
@IGx89 exactly - all the fixes are going to 3.1.
